### ISSUE TITLE
Eval bug: Gemma 4 26B A4B (QAT) full-GPU offload corrupts output on Vulkan (Radeon 890M gfx1150) - isolated to fused MMVQ kernel path

### ISSUE THREAD

**Issue by brostudiodev** [ggml-org/llama.cpp#27007]
### Name and Version

```
version: 10405 (e79e4bf66)   [also tested: master 0d0bfcd4, b10386]
built with GNU 11.4.0 for Linux x86_64
```

### Operating systems

Linux (Ubuntu 26.04 LTS "resolute")

### GGML backends

Vulkan

### Hardware

- AMD Ryzen AI 370 (Strix Point) — **Radeon 890M** iGPU (RDNA 3.5, **gfx1150**, 16 CU)
- RADV STRIX1, Mesa **26.0.3-1ubuntu1** (Vulkan 1.4.335) — also tested Mesa **26.1.6** (kisak PPA), identical result
- UMA device (`uma: 1`), 64 GB unified memory

### Models

- `ggml-org/gemma-4-26B-A4B-it-qat-q4_0.gguf` (target)
- `mtp-gemma-4-26B-A4B-it.gguf` (MTP drafter, separate file)

### Problem description & steps to reproduce

With **full-GPU offload** (`--n-gpu-layers 99`, no `--n-cpu-moe`), Gemma 4 26B A4B QAT produces **deterministically corrupted output** on this iGPU. Same for the non-QAT `gemma-4-26B-A4B-it-UD-Q5_K_M` and `Q4_K_M` quantizations.

Corruption is **byte-identical across runs** (temp 0) and takes one of two forms depending on config:
- **full repetition loops** — e.g. an entire answer replaced by a repeated word fragment (`"forma-se"`, `"0.5 stars"`, `"de-escalation"`), or
- **garbled prefix** then self-healed correct content — the classic signature e.g. `"It is possible to possible.thought\n<|channel>thought\n<channel|>The capital of France is **Paris**."`

Repro (standalone `llama-server`, temp 0, 4 prompts):
```
llama-server --model gemma-4-26B-A4B-it-qat-q4_0.gguf --n-gpu-layers 99 \
  --ctx-size 32768 --cache-type-k q8_0 --cache-type-v q8_0 --flash-attn on \
  --model-draft mtp-gemma-4-26B-A4B-it.gguf --spec-type draft-mtp \
  --spec-draft-n-max 2 --spec-draft-ngl 99 --threads 24 --ubatch-size 512
```

**Known-good baseline:** the same build/model with `--n-cpu-moe 99` (MoE experts on CPU) is byte-correct. So this is a Vulkan decode-path bug, not a model/data issue.

### Diagnostic findings (23-run env sweep, temp 0, adjudicated byte-vs-known-good + loop detection)

| Env / flag | Result |
|---|---|
| `GGML_VK_DISABLE_MMVQ=1` | **Strongest lead.** Full loops gone; only deterministic garbled prefix on the first prompt (other 3 byte-clean). Corruption lives in the **fused `mul_mat_vec_q` (MMVQ) kernel path**. |
| `GGML_VK_DISABLE_FUSION=1` | France prompt fixed, but the loop **moved** to another prompt (sky) — fusion-dependent. |
| `GGML_VK_DISABLE_MMVQ=1` + `DISABLE_COOPMAT` / `DISABLE_INTEGER_DOT_PRODUCT` / ubatch 256 / no drafter | Same as plain `DISABLE_MMVQ`. |
| `MAX_NODES_PER_SUBMIT=1`, `DISABLE_ASYNC`, `SERIALIZE_SUBMISSIONS`, `RADV_DEBUG=syncshaders`, `DISABLE_MULTI_ADD`, `DISABLE_GRAPH_OPTIMIZE`, `FORCE_MMVQ`, `ALLOW_SYSMEM_FALLBACK`, `DISABLE_HOST_VISIBLE_VIDMEM` | Identical corruption to baseline. |
| f16 KV cache + `DISABLE_MMVQ` | Worse: new French repetition loop appears. |
| `--flash-attn off` + MTP drafter | Load fails (drafter ctx init) — unrelated. |
| `GGML_VK_PREFER_HOST_MEMORY=1` | Load fails `vk::Queue::submit: ErrorDeviceLost` — unrelated. |

**Mesa attribution:** corruption is **byte-identical on Mesa 26.0.3 and 26.1.6** (both baseline loop and `DISABLE_MMVQ` garbled prefix). A RADV compiler miscompile would very likely perturb the corrupted output across a driver change; byte-identical output across two driver versions points to deterministic backend-side kernel selection / buffer layout / fusion logic rather than a driver shader bug.

**Build attribution:** same corruption on b10386, b10405, and current master (`0d0bfcd4`, 7 commits past b10405, none touching `ggml-vulkan`).

### Related upstream

- #24311 — same class on **Gemma 4 12B QAT** Vulkan (`unused49` loop), stale-closed. Confirms it is not specific to 26B A4B.
- #25777 — same class on **Intel Arc A770**: Gemma 4 26B A4B + Vulkan corruption/assert; report there shows the crash point moves with the memory path and is sensitive to `DISABLE_HOST_VISIBLE_VIDMEM` / ubatch — same latent-fusion/memory-path family.
- #24966 (open PR) and #25666 (open PR) — MMVQ on AMD UMA / speculative-decode steps; #25666 explicitly documents that MMVQ **"perturbs the logits"** in the small-`n` decode path, which is consistent with our `DISABLE_MMVQ` result.
- #21566 (CUDA, merged) and #19768 (Vulkan, merged) — GEMV-fusion buffer-overlap checks; both are already inside b10405 yet our decode path still corrupts, suggesting an unported `ggml_nrows=1`-style gap (the case #21566 had to extend its overlap check for on CUDA).

### Expected behavior

`--n-gpu-layers 99` on Vulkan should produce byte-identical output to the `n-cpu-moe` config at temp 0. It does not, and no runtime toggle or server flag produces correct output.

### Logs / additional context

Full raw results + adjudication: 28 sweep JSONs with byte-parity reference at `automationbro-infra/ai-pc-benchmarks/results/gemma4/sweep-fullgpu/` in github.com/brostudiodev/autonomous-living-infrastructure (SUMMARY.json has the per-toggle table). Happy to attach individual logs on request.

Full-GPU is ~38 t/s vs ~19 t/s with the CPU-MoE workaround on this APU, so this is worth fixing.

**Comment by brostudiodev** [comment]
Following up on the two related open PRs: **#24966** (disable MMVQ on AMD UMA) and **#25666** (do not enable MMVQ for speculative-decode steps on AMD). My sweep independently confirms the same root cause from the correctness side: with `GGML_VK_DISABLE_MMVQ=1` the full repetition loops disappear entirely and only a deterministic garbled prefix remains (self-heals to correct content), so the corruption lives in the fused `mul_mat_vec_q` path during decode. Happy to test a build containing either PR on the 890M (gfx1150) if that helps.

**Comment by Zaneboy** [comment]
_Disclaimer: this post has been created by ChatGPT 5.6 Sol after extensive testing and debugging because without it I'd have given up long ago xD
I hope this helps._

[mmvq-default.txt](https://github.com/user-attachments/files/31104286/mmvq-default.txt)
[mmvq-disabled.txt](https://github.com/user-attachments/files/31104285/mmvq-disabled.txt)
[mmvq-forced.txt](https://github.com/user-attachments/files/31104284/mmvq-forced.txt)
[qwen3-8b-q4km-ops.txt](https://github.com/user-attachments/files/31104283/qwen3-8b-q4km-ops.txt)

I can reproduce what appears to be the same Vulkan MMVQ correctness issue on a discrete RDNA4 GPU, with a standalone `test-backend-ops` reproduction.

### Environment

- Windows 11
- AMD Radeon RX 9070 XT 16 GB (RDNA4 / gfx1201)
- AMD Adrenalin 26.6.4
- Windows driver: `32.0.31021.5001`
- Vulkan: AMD proprietary driver, API 1.4.349
- llama.cpp current master:
  - build 10446
  - commit `adb55e514`
  - MSVC 19.44.35228.0
- Also reproduced corrupted inference with official Vulkan builds:
  - b10048 (`2e1fd7649`)
  - b10068 (`571d0d540`)

### Inference behavior

Qwen3-8B:

| Quant | Default Vulkan | `GGML_VK_DISABLE_MMVQ=1` |
|---|---|---|
| Q4_K_M | corrupted | correct |
| Q5_0 | corrupted | correct |
| Q6_K | correct | — |
| Q8_0 | correct | — |

Additionally:

- `GGML_VK_FORCE_MMVQ=1` makes normally-correct Q6_K corrupt.
- `GGML_VK_FORCE_MMVQ=1` makes normally-correct Q8_0 corrupt.
- `GGML_VK_DISABLE_INTEGER_DOT_PRODUCT=1` also fixes Q4/Q5, but disabling MMVQ alone is sufficient.
- Disabling CoopMat, CoopMat2, FP16, and Flash Attention did not fix it.

### Standalone backend reproduction

I generated actual Qwen3-8B Q4_K_M graph operations:

```powershell
.\test-export-graph-ops.exe `
    -hf Qwen/Qwen3-8B-GGUF:Q4_K_M `
    -o .\qwen3-8b-q4km-ops.txt
```

Then:

```powershell
.\test-backend-ops.exe test `
    -b Vulkan0 `
    -o MUL_MAT `
    --test-file .\qwen3-8b-q4km-ops.txt
```

#### Default dispatch

Result: **10/14 pass**

All failures are Q4_K, n=1:

```text
ERR = 1.000074547
sources=q4_K[4096,1024,1,1],f32[4096,1,1,1]

ERR = 1.000042645
sources=q4_K[4096,4096,1,1],f32[4096,1,1,1]

ERR = 1.000027334
sources=q4_K[12288,4096,1,1],f32[12288,1,1,1]

ERR = 0.999935807
sources=q4_K[4096,12288,1,1],f32[4096,1,1,1]
```

Equivalent n=512 operations pass.

#### `GGML_VK_DISABLE_MMVQ=1`

Same exported test file:

```text
14/14 tests passed
Backend Vulkan0: OK
```

#### `GGML_VK_FORCE_MMVQ=1`

Result: **7/14 pass**

Now both Q4_K and Q6_K n=1 operations fail:

```text
ERR = 0.999943593
sources=q4_K[4096,1024,1,1],f32[4096,1,1,1]

ERR = 0.999723485
sources=q6_K[4096,1024,1,1],f32[4096,1,1,1]

ERR = 0.999945831
sources=q4_K[4096,4096,1,1],f32[4096,1,1,1]

ERR = 0.999987618
sources=q4_K[12288,4096,1,1],f32[12288,1,1,1]

ERR = 0.998667264
sources=q6_K[12288,4096,1,1],f32[12288,1,1,1]

ERR = 1.000006301
sources=q4_K[4096,12288,1,1],f32[4096,1,1,1]

ERR = 0.992453083
sources=q6_K[4096,151936,1,1],f32[4096,1,1,1]
```

Again, the equivalent n=512 operations pass.

### Conclusion

The numerical failure follows Vulkan MMVQ selection:

- disable MMVQ -> all exported operations pass
- force MMVQ -> quant types that normally work also fail
- large-N / matrix-matrix operations pass
- small-N / matvec MMVQ operations fail with error ~1.0

This looks like a correctness problem in the Vulkan small-N MMVQ / `mul_mat_vec_q` path rather than model data, quantization generally, or Vulkan generally.

### Related

- #27007 - AMD Vulkan corruption independently isolated to MMVQ on Radeon 890M
- #21151 - Q4_K/Q5_K MMVQ routing on AMD RDNA3
- #24966 - MMVQ handling on AMD UMA
- #25666 - MMVQ perturbing logits during AMD speculative decode

### Workaround

```text
GGML_VK_DISABLE_MMVQ=1
```

This restores correct inference for Q4_K_M and Q5_0 on this system.

Attached are the exported `qwen3-8b-q4km-ops.txt` graph operations and the complete `mmvq-default.txt`, `mmvq-disabled.txt`, and `mmvq-forced.txt` test-backend-ops logs used for the results above.
